x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,340 @@
+                    } else {
+                        context.addIgnoredField(fieldType.name());
+                        continue;
+                    }
+                }
+
+                // Make sure that the value is a number. Probably this will change when
+                // new aggregate metric types are added (histogram, cardinality etc)
+                if (e.getValue() instanceof Number == false) {
+                    if (ignoreMalformed.value() == false) {
+                        throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                            "] of field [" + fieldType.name() + "] must be a number");
+                    } else {
+                        context.addIgnoredField(fieldType.name());
+                        continue;
+                    }
+                }
+
+                // Make sure that value_count is not a negative value
+                if (Metrics.value_count.toString().equals(metricName) == true) {
+                    if (e.getValue() instanceof Integer == false) {
+                        if (ignoreMalformed.value() == false) {
+                            throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                                "] of field [" + fieldType.name() + "] must be an integer number");
+                        } else {
+                            context.addIgnoredField(fieldType.name());
+                            continue;
+                        }
+                    }
+
+                    Integer metricValue = (Integer) e.getValue();
+                    if (metricValue.intValue() < 0) {
+                        if (ignoreMalformed.value() == false) {
+                            throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                                "] of field [" + fieldType.name() + "] must not be a negative number");
+                        } else {
+                            context.addIgnoredField(fieldType.name());
+                            continue;
+                        }
+                    }
+                }
+
+                parsedMetricsMap.put(metricName, e.getValue());
+            }
+        } else {
+            throw new MapperParsingException("Cannot parse aggregate metric for field [{}]",
+                null,fieldType.name());
+        }
+
+        // TODO: Implement null_value support?
+        if (parsedMetricsMap.keySet().containsAll(metrics.value()) == false
+            && ignoreMalformed.value() == false) {
+            throw new IllegalArgumentException("Aggregate metric field [" + fieldType.name() +
+                "] must contain all metrics " + metrics.value().toString());
+        }
+
+        if (fieldType().hasDocValues()) {
+            BytesStreamOutput streamOutput = new BytesStreamOutput();
+            streamOutput.writeMap(parsedMetricsMap);
```

Comment:
Question to @iverase and @jimczi : 

Is this method of serializing the `parsedMetricsMap` preferred as opposed to 
```
            for (Map.Entry<String, Object> e : parsedMetricsMap.entrySet()) {
                streamOutput.writeString(e.getKey());

                Double d = null;
                if (e.getValue() instanceof Double) {
                    d = (Double) e.getValue();
                } else if (e.getValue() instanceof Integer) {
                    d = Double.valueOf((Integer) e.getValue());
                }
                streamOutput.writeLong(Double.doubleToRawLongBits(d));

            }
````
We should take into consideration: 

- How it's best to deserialize this map.
- Values are now Double objects, but we may want to serialize histrograms or cardinality structures.

Reply:
I think we said that we should give each metric his dedicated sub-field ? In such case you can create one docvalue field for each metric.

Reply:
I actually prefer this way and I would recommend to use `ByteBuffersDataOutput` from Lucene instead of BytesStreamOutput. The idea behind is that when deserialising you can use `ByteArrayDataInput` which it can be reused and you some allocations per document.

See Histogram field type for reference. wdyt?